### ISSUE TITLE
Search on lib.haxe.org could be better

### ISSUE THREAD

**Issue by explorigin** [HaxeFoundation/haxelib#169]
tl;dr; Need full-text search and some sort of ranking system.

I was having trouble using the std jquery extern so I was looking for an alternative.  I search "jquery" and I get http://lib.haxe.org/p/jquery.

"OK cool, that looks more up-to-date."

Later I learn that http://lib.haxe.org/p/jQueryExtern is the really up-to-date one that everyone is using.  The problem is that a search of "jquery" doesn't reveal it.  Even if it did, as a new developer, I would have no idea which one to choose.  We should have a ranking system to give a relative idea of which ones are more used.  NPMjs uses number of downloads which seems pretty good to me.

**Comment by Simn** [comment]
I agree, but... just wait a little bit longer. :)

**Comment by jasononeil** [comment]
New haxelib website is in progress, and should be done in a few weeks.  We're aiming to release it along with the Haxe 3.2 release.

All the project lists on the new site are sorted by popularity / downloads, so in this case the most popular jQuery library would float to the top.  I'm also getting rid of the logic on the old site that automatically skipped the search page if a library had the exact same name as the search term.

Having said that, I'm not sure if doing our own search is the best solution.  I think perhaps a general Google search of the site could pick up not just keywords in the title / description, but also in READMEs, documentation etc.

@Simn  / @ncannasse, any preference for the haxelib search feature?  Google or basic title/description, sorted by popularity?

**Comment by jdonaldson** [comment]
:clap:

**Comment by Simn** [comment]
This can be considered resolved with the new website.

**Comment by jasononeil** [comment]
The new search no longer automatically goes to the library page, but shows a list of projects, which includes a download count.

But it's pretty amateur still... the results are sorted alphabetically, not by popularity or relevance etc.  I'm not sure what the best approach would be.  Luckily any given search doesn't have too many results so it's manageable at the moment.

If anyone has some concrete suggestions I'd be happy to tackle them...

**Comment by ibilon** [comment]
A simple method to rank results by relevance is the tf-idf method: https://en.wikipedia.org/wiki/tf-idf

The tf is easy to calculate and if we only stock the denominator of the idf it's easy to update when new versions/libraries are submitted.

Storing the index could be an issue though, no idea how much that'd be, less than the corpus size for sure, which is the descriptions, tags and maybe readmes?

On the positive side it's really easy to store, it's "just" a lot of key value pairs, and should be really fast to retrieve and compute.

**Comment by explorigin** [comment]
Just make the columns sortable.  You'll go a long way without having to introduce a new ranking method.
